src/poll.rs

```diff
@@ -145,3 +270,339 @@ pub fn selector(poll: &Poll) -> &sys::Selector {
 pub fn selector_mut(poll: &mut Poll) -> &mut sys::Selector {
     &mut poll.selector
 }
+
+/*
+ *
+ * ===== Registration =====
+ *
+ */
+
+impl Registration {
+    pub fn new(poll: &Poll, token: Token, interest: EventSet, opts: PollOpt) -> Registration {
+        let queue = poll.readiness_queue.clone();
+        let node = queue.new_readiness_node(token, interest, opts);
+
+        Registration {
+            node: node,
+            queue: queue,
+        }
+    }
+
+    pub fn update(&self, poll: &Poll, interest: EventSet, opts: PollOpt) -> io::Result<()> {
+        if !self.queue.identical(&poll.readiness_queue) {
+            return Err(io::Error::new(io::ErrorKind::Other, "nope"));
+        }
+
+        self.node().interest.set(interest);
+        self.node().opts.set(opts);
+
+        // If the node is currently ready, re-queue?
+        if !event::is_empty(self.readiness()) {
+            self.queue_for_processing(None);
+        }
+
+        Ok(())
+    }
+
+    pub fn readiness(&self) -> EventSet {
+        event::from_usize(self.node().events.load(Ordering::Relaxed))
```

Comment:
I would personally **very strongly recommend** using `SeqCst` everywhere. It's far easier to selectively relax orderings, and profiling can be used to figure out which ones are on the hot paths.

If everything is Relaxed or Acquire/Release it's incredibly difficult to reason about whether the orderings are actually correct. They're all basically related to global decisions, so you have to understand basically everything to reason about any local ordering as well. Thankfully if most orderings are `SeqCst` you can safely ignore most of the program as they'll work "as expected".